Login

JrSchoolAccount opened the pull request:
# Frontend

## Login Page /login

- Error handling and validation
- Login logic

## SignUp Page /registrera

- Error handling and validation
- Registration logic

## Log Out Button
- Simple button to call api/logout

## Profile page
- Simple profile page with logout button

---

# Backend

- Session (Cookies and JWT)
- Sign Up API /api/signup
- Login API /api/login
- Logout API /api/logout
- Middleware to protect /profile

---

# MongoDB

## Users Collection
```json
{
  "_id": {"$oid": "66349ec24d0e094813a01d94"},
  "name": "John Doe",
  "email": "john@example.com",
  "password": "$2b$10$PV9m4WELfhvFdceIONrGbeIeDoGqJr4Gwha1dfk059zao7eH5YZ9e",
  "__v": 0
}

```
---

# Cypress E2E Test
- Register new user test
- Login user test
---

# Screenshots

### Sign up
<img width="639" alt="Screenshot 2024-05-16 at 08 39 18" src="https://github.com/JrSchoolAccount/Kino-site-nextjs/assets/144328779/19f538c5-e7c5-48b5-9093-d85c50efe093">

---

### Login
<img width="641" alt="Screenshot 2024-05-16 at 08 38 17" src="https://github.com/JrSchoolAccount/Kino-site-nextjs/assets/144328779/d52356c9-8110-4ad4-b561-5d7b8028ef1b">

---

NelsonTheBrave commented on app/layout.tsx:
```diff
@@ -1,3 +1,4 @@
+import { useState } from 'react';
```
Why is useState imported here? It doesn't seem to be used in this file?

JrSchoolAccount replied:
i removed it.

NelsonTheBrave commented on app/ui/signup.tsx:
```diff
@@ -0,0 +1,141 @@
+            <Grid item xs={12}>
+              <TextField
+                autoComplete="given-name"
+                name="name"
```
There is something a bit weird with how the name-field works. When loading the signup page it is automatically highlighted, but the placeholder text doesn't show, which means you don't understand you're supposed to write a firstname and lastname unless you click somewhere else first, or write and then delete the text. I guess it's the autofocus property that does this, don't know if there is a way to have autofocus and also show placeholder at the same time

NelsonTheBrave commented on cypress/e2e/loginUser.cy.ts:
```diff
@@ -0,0 +1,17 @@
+    cy.get('button[type="submit"]').click();
+    cy.wait(2000);
+
+    cy.url().should('include', '/profil');
```
This doesn't work for me... I mean the actual functionality to redirect to /profil doesn't happen when I log in :O I go back to '/'

JrSchoolAccount replied:
Changed :)

NelsonTheBrave reviewed (changes requested):
Really nice work! Good bunch of code you have in this PR :) 
I can't really comment on the quality of the login functionality, encryption etc since I haven't worked with it myself, but it seems like you have used both hashing and salt for the password checking process which is nice.

I have noted a few issues that I found. The main one I think is that after logging in I neither get redirected to the profile page, nor does the navBar change to represent that the session has started. It seems like that is working for you locally, as I've understood, so I'm not sure what would be the reason.

Another thing that would be nice, but not a must, is if there was some kind of message when you register a new user, and not just redirecting to the startpage. I felt the user experience was a bit uncertain whether something actually happened or not.

I would also put just a little bit of margin between the two buttons in the navbar (logga in and registrera) to make it look nicer :)

Also I realised we should talk in the group about where we put or models, in separate files or in one collected file like we do with the types in definitions.ts. But that is more a general thing than relating to your branch.

Great job!

FelixHultman commented:
I've look at the code but im not really sure i can give you any good feedback since i dont really understand the code that good, but as Erik said i see you are using alot of the elements we looked at in class, so thats cool. 

Im having kinda the same problem as Erik when im trying out the branch, so i guess it will be easiest if we talk about it the whole group maybe? but it is impressive work, looking forward to you presentation :)

FelixHultman commented on app/ui/navbar.tsx:
```diff
@@ -20,31 +18,39 @@ import { Stack } from '@mui/material';
 
 const links = [
   { name: 'Om oss', href: '/om-oss' },
   { name: 'Aktuella visningar', href: '/aktuella-vinsningar' },
```
this should

FelixHultman replied:
this should be aktuella-visningar! :)

JrSchoolAccount reviewed (commented):
Removed navbar login/logout and moved to a profile page solution instead


https://github.com/JrSchoolAccount/Kino-site-nextjs/assets/144328779/21d8430b-e03e-4ccb-bff7-00d6e5332c0d

simonpizevski reviewed (changes requested):
I think that it works very good and the new changes to the profile page works good too!. Great job. 

Im not the best at encryption but good job with the hashed passwords and user creation. 

I think that instead of "mina bokningar" we could name it to "min profil" or something like that.

JrSchoolAccount commented:
I think i have updated the pr and addressed all the issues found :)

Ready for a final review!

FelixHultman commented:
> I think i have updated the pr and addressed all the issues found :)
> 
> Ready for a final review!

i cant seem to login, (failed to verify session). 
another small thing is that you cant use å, ä, ö when register a new member( not that important just a note)

JrSchoolAccount commented:
> > I think i have updated the pr and addressed all the issues found :)
> > Ready for a final review!
> 
> i cant seem to login, (failed to verify session). another small thing is that you cant use å, ä, ö when register a new member( not that important just a note)

@FelixHultman 
I fixed the "åäö" issue :) 
Dont have a login error on my end. do you mind trying again

FelixHultman commented:
> > > I think i have updated the pr and addressed all the issues found :)
> > > Ready for a final review!
> > 
> > 
> > i cant seem to login, (failed to verify session). another small thing is that you cant use å, ä, ö when register a new member( not that important just a note)
> 
> @FelixHultman I fixed the "åäö" issue :) Dont have a login error on my end. do you mind trying again

i just get failed to verify session when trying to log in, but it might be something else im doing wrong. Lets discuss it at the meeting.

NelsonTheBrave reviewed (approved):
Now it works to log in and access the profile page! Just check spelling of "Mina profil" in the navbar ;)

simonpizevski reviewed (approved):
Works great :)
